### PR TITLE
MIM-277 + MIM-296 resources panel

### PR DESCRIPTION
**Pull request by SandraMonnier**
https://albasim.atlassian.net/browse/MIM-277
https://albasim.atlassian.net/browse/MIM-296

https://docs.google.com/document/d/1PLhFgMfWTVO7rIFbF1AFdgaL98yCUuH7pVzANn5dGeU/edit#heading=h.su8td74siwbc

### CODE REVIEW

File: gameModel/libs/ClientScript/game/common/actions/actionBase.ts [hunk 1]
```diff
@@ -1202,42 +1203,70 @@ export class MoveResourcesAssignTaskAction extends StartEndAction {
...
+  }
+
+  public getMessage(): string {
+    let arg0: string = '';
+    for (const res in this.sentResources) {
+      arg0 += this.sentResources[res as ResourceType];
+      arg0 += ' ';
+      arg0 += getTranslation('mainSim-resources', '' + res);
+      arg0 += ', ';
+    }
```
Comment (xgo-work) [inline gameModel/libs/ClientScript/game/common/actions/actionBase.ts:1249]:
```suggestion
      const arg0 = this.sentResources.map( res => this.sentResources[res as ResourceType] + ' ' + getTranslation('mainSim-resources', '' + res)).join(', ');
```

File: gameModel/libs/ClientScript/game/common/actions/actionBase.ts [hunk 2]
```diff
@@ -1202,42 +1203,70 @@ export class MoveResourcesAssignTaskAction extends StartEndAction {
...
+      arg0,
+      arg1,
+      arg2,
+      arg3,
+      arg4,
+    ]);
+  }
+
+  public getEmitter(): string {
+    return getCurrentState().getActorById(this.ownerId)!.Role as unknown as TranslationKey;
```
Comment (xgo-work) [inline gameModel/libs/ClientScript/game/common/actions/actionBase.ts:1264]:
```suggestion
    return getCurrentState().getActorById(this.ownerId)?.Role as unknown as TranslationKey;
```

File: gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  Radio = 'Radio', // by radio
+}
+
+/**
+ * Which communication media can be used to reach a resource
+ */
+enum AvailabilityCommMedia {
+  None = 'None',
+  Direct = 'Direct',
+  Radio = 'Radio',
```
Comment (xgo-work) [inline gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts:18]:
```suggestion
  Direct = CommMedia.Direct,
  Radio = CommMedia.Radio,
```

File: gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  Both = 'Both',
+}
+
```
Comment (xgo-work) [inline gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts:21]:
```suggestion
enum AvailabilityCommMedia {
  Direct = CommMedia.Direct,
  Radio = CommMedia.Radio,
  Both = 'Both',
}

type Availability = `${AvailabilityCommMedia}` | undefined;

```

File: gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+/**
+ * Define how a resource at some location doing some task can be reached
+ */
+const communicationMatrix: Record<
+  LOCATION_ENUM,
+  Partial<Record<TaskType, AvailabilityCommMedia>>
```
Comment (xgo-work) [inline gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts:27]:
```suggestion
  Partial<Record<TaskType, Availability>>
```

File: gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      );
+    } else if (commMedia === CommMedia.Radio) {
+      return (
+        commMediaAvailable === AvailabilityCommMedia.Radio ||
+        commMediaAvailable === AvailabilityCommMedia.Both
+      );
+    }
+  }
+
+  return false;
```
Comment (xgo-work) [inline gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts:90]:
```suggestion
  const commMediaAvailable: Availability = communicationMatrix[location][taskType];
  return commMediaAvailable === 'Both' || commMediaAvailable === commMedia;
```

File: gameModel/libs/ClientScript/game/common/simulationState/taskStateAccess.ts [hunk 7]
```diff
@@ -24,24 +25,47 @@ export function getAllTasks(state: Readonly<MainSimulationState>): Readonly<Task
 }
 
 /**
- * Temporary filtering implementation of local communication restrictions on incident site
+ * Fetch the tasks whose resources can be reached by an actor, at a location, via a communication media
```
Comment (xgo-work) [inline gameModel/libs/ClientScript/game/common/simulationState/taskStateAccess.ts:28]:
```suggestion
 * Fetch the tasks which resources can be reached by an actor, at a location, via a communication media
```

File: gameModel/libs/ClientScript/gameInterface/actionsButtonLogic.ts [hunk 8]
```diff
@@ -145,9 +149,27 @@ function fetchMoveResourcesAssignTaskValues() {
   };
 
   // Reset interfaceState
-  const newState = Helpers.cloneDeep(Context.interfaceState.state);
-  newState.resources[paramKey] = getEmptyFunc();
-  Context.interfaceState.setState(newState);
+  if (panel === SelectedPanel.resources) {
```
Comment (xgo-work) [inline gameModel/libs/ClientScript/gameInterface/actionsButtonLogic.ts:152]: Pretty esthetics again but I would take out the `deepClone` before the `if` and the `Context.interfaceState.setState(newState);` after. just to have those once only

File: gameModel/libs/ClientScript/game/common/simulationState/taskStateAccess.ts [hunk 9]
```diff
@@ -24,24 +25,47 @@ export function getAllTasks(state: Readonly<MainSimulationState>): Readonly<Task
...
+
+export function isReachable(
+  state: Readonly<MainSimulationState>,
+  actorId: ActorId | undefined,
+  location: LOCATION_ENUM,
+  taskId: TaskId,
+  commMedia: CommMedia
+): boolean {
+  const task: TaskBase = internallyGetTask(state, taskId);
+  const actor: Readonly<Actor> | undefined = actorId ? state.getActorById(actorId) : undefined;
```
[the remark this answers is not in the document]
Comment (SandraMonnier) [reply gameModel/libs/ClientScript/game/common/simulationState/taskStateAccess.ts:49]: I tried, but there is a compilation error as getActorById does not deal with undefined

File: gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts (hunk 4 above)
Comment (SandraMonnier) [reply gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts:21]: `Direct = CommMedia.Direct`
Does not compile if the second part is something computed.

File: gameModel/libs/ClientScript/game/common/actions/actionBase.ts (hunk 1 above)
Comment (SandraMonnier) [reply gameModel/libs/ClientScript/game/common/actions/actionBase.ts]: Done in commit b87082b5

File: gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts (hunk 5 above)
Comment (SandraMonnier) [reply gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts]: Done in commit b87082b5

File: gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts (hunk 6 above)
Comment (SandraMonnier) [reply gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts]: Done in commit b87082b5

File: gameModel/libs/ClientScript/gameInterface/actionsButtonLogic.ts (hunk 8 above)
Comment (SandraMonnier) [reply gameModel/libs/ClientScript/gameInterface/actionsButtonLogic.ts:153]: Done in commit b87082b5

File: gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts (hunk 4 above)
Comment (xgo-work) [reply gameModel/libs/ClientScript/game/common/resources/resourceReachLogic.ts]: How sad -_-

File: gameModel/libs/ClientScript/game/common/simulationState/taskStateAccess.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (xgo-work) [reply gameModel/libs/ClientScript/game/common/simulationState/taskStateAccess.ts:49]: Yeah right, then I guess we can add the undefined typing in the function's signature

**Pull request merged**